### PR TITLE
Added common dir and Makefile

### PR DESCRIPTION
**Pull request by muhmuz**
- Added common dir
- Added four files in common dir
- Added tests directory with correctness and performance directories
- Added source dir
- Added Makefile

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -178,11 +179,12 @@ endif
 CFLAGS2 := ${CFLAGS_PROJECT} ${CFLAG_ENTROPY} ${CFLAGS} ${CFLAGS_V}
 
 # Run and output directory
-SUBD ?= default
 RUN_DIR ?=
 ifeq (,$(RUN_DIR))
   RUN_DIR_FLAG = 0
-  RUN_DIR := ${ENV_DIR}/RUN/${SUBD}
+  RUN_DIR := ${TESTS_DIR}/RUN
```
Comment (mmhus) [inline Makefile]: It should be `RUN_DIR := ${ENV_DIR}/RUN/${SUBD}`

File: tests/correctness/program.S [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# ----------------------------------------------------------------
```
Comment (mmhus) [inline tests/correctness/program.S]: These tests should be in their own separate folders.

File: Makefile [hunk 3]
```diff
@@ -196,8 +198,21 @@ LIB_DIR ?=
 ##############################################################################
 # Internal variables
 
-ELF_FILE := ${RUN_DIR}/test.elf
-DIS_FILE := ${RUN_DIR}/test.disasm
+# Output binary and disassembly of c source file
```
Comment (mmhus) [inline Makefile]: Please redo, as we discussed, this is not correct.

Comment (mmhus) [reply Makefile]: Change it to how it was before, and just add the spike and qemu output folders

File: Makefile [hunk 4]
```diff
@@ -250,8 +265,11 @@ endif
 ##############################################################################
 # Expansions
 
-COMPILE_EXP = "riscv64-unknown-elf-gcc ${FRAMEWORK_CFLAGS} ${CARCH} ${COPT} ${CFLAGS2} ${FRAMEWORK_SRCS} ${COMMON_SRCS} ${ENV_SRCS} ${LIB_SRCS} ${LDFLAGS} -o $@"
-DISM_EXP = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
+COMPILE_EXP_C = "riscv64-unknown-elf-gcc ${FRAMEWORK_CFLAGS} ${CARCH} ${COPT} ${CFLAGS2} ${FRAMEWORK_SRCS} ${COMMON_SRCS} ${ENV_SRCS} ${LIB_SRCS} ${LDFLAGS} ${TESTS_DIR}/${C_SRC} -o $@"
```
Comment (mmhus) [inline Makefile]: Why is this change needed?

File: Makefile [hunk 5]
```diff
@@ -250,8 +265,11 @@ endif
...
+DISM_EXP_C = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
```
Comment (mmhus) [inline Makefile]: same as above, why?

File: Makefile [hunk 6]
```diff
@@ -263,48 +281,82 @@ DISM_EXP = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
 default: compile spike
 
 setup: test
-	mkdir -p ${RUN_DIR}
+	mkdir -p ${SPIKE_DIR}
```
Comment (mmhus) [inline Makefile]: This should not replace RUN_DIR, but add two folders for qemu and spike. It should also never be named something like SPIKE_DIR, if its function is not just to add a spike directory.

File: Makefile [hunk 7]
```diff
@@ -263,48 +281,82 @@ DISM_EXP = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
...
 
-${ELF_FILE}: setup ${ENV_SRCS}
+# Compile .c Source File
+${ELF_C_FILE}: setup ${TESTS_DIR}/${C_SRC}
 	@echo ""
-	@echo ${COMPILE_EXP} > ${RUN_DIR}/compile_cmd_rerun.sh
-	@chmod u+x ${RUN_DIR}/compile_cmd_rerun.sh
-	${RISCV_PREFIX}-gcc ${FRAMEWORK_CFLAGS} ${CARCH} ${COPT} ${CFLAGS2} ${FRAMEWORK_SRCS} ${COMMON_SRCS} ${ENV_SRCS} ${LIB_SRCS} ${LDFLAGS} -o $@
+	@echo ${COMPILE_EXP_C} > ${SPIKE_DIR}/c_compile_cmd_rerun.sh
```
Comment (mmhus) [inline Makefile]: pls fix these and below to be what they were before.

File: Makefile [hunk 8]
```diff
@@ -263,48 +281,82 @@ DISM_EXP = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
...
-	@echo "export DBG=${ELF_FILE}" > ${RUN_DIR}/temp_gdb_export.sh
-	@echo "${RISCV_PREFIX}-gdb --exec=${realpath ${ELF_FILE}} --symbols=${realpath ${ELF_FILE}}" >> ${RUN_DIR}/temp_gdb_export.sh
+	@echo "export DBG=${ELF_C_FILE}" > ${SPIKE_DIR}/c_temp_gdb_export.sh
+	@echo "${RISCV_PREFIX}-gdb --exec=${realpath ${ELF_C_FILE}} --symbols=${realpath ${ELF_C_FILE}}" >> ${SPIKE_DIR}/c_temp_gdb_export.sh
 
-compile: ${ELF_FILE} ${DIS_FILE}
+compile_c: ${ELF_C_FILE} ${DIS_C_FILE}
 	@echo ""
 
+# Compile .S Source File
```
Comment (mmhus) [inline Makefile]: unnecessary code.

File: Makefile [hunk 9]
```diff
@@ -263,48 +281,82 @@ DISM_EXP = "riscv64-unknown-elf-objdump ${OBJDUMP_FLAGS} $< > $@"
...
+ifndef C_SRC
+	$(error C_SRC is undefined)
 endif
 
-spike:
+# Run C Binary on Spike
+spike_c:
 	@echo ""
-	$(SPIKE) --isa=$(SPIKE_ISA) -l --log-commits ${PK} ${ELF_FILE} 1> ${RUN_DIR}/$@.out 2> ${RUN_DIR}/$@.err
+	$(SPIKE) --isa=$(SPIKE_ISA) -l --log-commits ${PK} ${ELF_C_FILE} 1> ${SPIKE_OUT_C} 2> ${SPIKE_LOG_C}
```
Comment (mmhus) [inline Makefile]: pls redo it the way it was and create separate test folders.

File: tests/correctness/program.S [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	li a0, 1	#To output the message on stdout
+	
+	# Load address of message into a1
+	la a1, message
+	
+	# Load length of message into a2
+	li a2, 14	#Length of "Hello, World!\n"
+	
+	# Load the syscall number for write into a7
+	li a7, 64	#syscall number for write
```
Comment (mmhus) [inline tests/correctness/program.S]: Please add reference for where you found this syscall, and is it even correct.

File: modulefiles/qemu-riscv64/5.0.0 [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#%Module1.0
```
Comment (mmhus) [inline modulefiles/qemu-riscv64/5.0.0:3]: missing copyright

File: modulefiles/qemu-riscv64/8.2.1 [hunk 12]
```diff
@@ -0,0 +1,27 @@
+#%Module1.0
```
Comment (mmhus) [inline modulefiles/qemu-riscv64/8.2.1:3]: missing copyright

File: modulefiles/riscv64-unknown-elf-gcc/13.2.0 [hunk 13]
```diff
@@ -0,0 +1,35 @@
+#%Module1.0
```
Comment (mmhus) [inline modulefiles/riscv64-unknown-elf-gcc/13.2.0:3]: missing copyright

File: modulefiles/riscv64-unknown-elf-gcc/14.1.1 [hunk 14]
```diff
@@ -0,0 +1,35 @@
+#%Module1.0
```
Comment (mmhus) [inline modulefiles/riscv64-unknown-elf-gcc/14.1.1:3]: missing copyright

File: tests/correctness/program.S (hunk 2 above)
Comment (mmhus) [inline tests/correctness/program.S]: missing copyright

File: modulefiles/qemu-riscv64/5.0.0 (hunk 11 above)
Comment (muhmuz) [reply modulefiles/qemu-riscv64/5.0.0:3]: added copyright

File: modulefiles/qemu-riscv64/8.2.1 (hunk 12 above)
Comment (muhmuz) [reply modulefiles/qemu-riscv64/8.2.1:3]: added copyright

File: modulefiles/riscv64-unknown-elf-gcc/13.2.0 (hunk 13 above)
Comment (muhmuz) [reply modulefiles/riscv64-unknown-elf-gcc/13.2.0:3]: added copyright

File: modulefiles/riscv64-unknown-elf-gcc/14.1.1 (hunk 14 above)
Comment (muhmuz) [reply modulefiles/riscv64-unknown-elf-gcc/14.1.1:3]: added copyright

File: Makefile (hunk 3 above)
Comment (muhmuz) [reply Makefile]: resolved it as per the comments

File: Makefile (hunk 7 above)
Comment (muhmuz) [reply Makefile]: done

File: Makefile (hunk 9 above)
Comment (muhmuz) [reply Makefile]: done

File: tests/correctness/program.S (hunk 10 above)
Comment (muhmuz) [reply tests/correctness/program.S]: Here's the reference of syscalls [syscalls reference](https://gpages.juszkiewicz.com.pl/syscalls-table/syscalls.html)

File: Makefile (hunk 1 above)
Comment (muhmuz) [reply Makefile]: Change it as per the comment

File: tests/correctness/program.S (hunk 2 above)
Comment (muhmuz) [reply tests/correctness/program.S]: Done

File: Makefile (hunk 4 above)
Comment (muhmuz) [reply Makefile]: reverted

**Review by mmhus — APPROVED** [review]
